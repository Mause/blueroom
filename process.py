import json
import sys
from asyncio import gather
from datetime import datetime, timedelta

import bs4
import httpx
import uvloop
from tqdm import tqdm


def groupby(iterable, key):
    groups = {}
    for item in iterable:
        groups.setdefault(key(item), []).append(item)
    return groups


def make_date(domain, date):
    start = datetime.fromisoformat(date["DateTime"])
    duration = timedelta(minutes=date["Runtime"])
    return {
        "start": start.isoformat(),
        "end": (start + duration).isoformat(),
        "venue": date["VenueName"],
        "url": f"https://tix.{domain}" + date["URL"],
    }


def get_event_url(domain: str, event: dict) -> str:
    path = event["URL"]
    if path.count("/") == 3:
        path = path.rsplit("/", 1)[0]
    return f"https://{domain}{path.lower()}/"


def boop(domain, shows, descriptions):
    for show, instances in shows.items():
        i = instances[0]
        meta = descriptions[show]
        yield {
            "item_hash": i["Hash"],
            "title": i["Name"],
            "url": get_event_url(domain, i),
            "html_desc": meta,
            "desc": i["DescriptionBrief"],
            "dates": [make_date(domain, instance) for instance in instances],
        }


async def get_show(
    domain: str, client: httpx.AsyncClient, key: str, event: dict
) -> tuple[str, str | None]:
    res = (
        await client.get(
            get_event_url(domain, event),
            headers={"User-Agent": "Mozilla/5.0"},
            follow_redirects=True,
        )
    ).text
    soup = bs4.BeautifulSoup(res, "html.parser")
    html_desc = soup.css.select_one(".event-desc")
    return (key, str(html_desc) if html_desc else None)


async def main(argv: list[str]) -> None:
    domain = argv[0] if argv else "blueroom.org.au"
    url = f"https://tix.{domain}/api/v1/Items/Browse"
    client = httpx.AsyncClient()
    data = (await client.get(url)).json()
    shows = groupby(data["Items"], lambda x: x["Name"].replace(" - Opening Night", ""))
    descriptions = dict(
        tqdm(
            await gather(
                *[
                    get_show(domain, client, key, values[0])
                    for key, values in shows.items()
                ]
            )
        )
    )
    with open(f"output/{domain}.json", "w") as f:
        json.dump(list(boop(domain, shows, descriptions)), f, indent=2)


if __name__ == "__main__":
    uvloop.run(main(sys.argv[1:]))
