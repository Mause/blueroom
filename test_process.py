from datetime import datetime, timezone

import uvloop
from httpx import Response
from respx import MockRouter
from syrupy.assertion import SnapshotAssertion

from process import process_domain


def test_main(respx_mock: MockRouter, snapshot: SnapshotAssertion) -> None:
    respx_mock.get("https://blueroom.org.au/event/1/").mock(
        Response(
            status_code=200,
            json={
                "Id": 1,
                "Name": "Test Event",
                "Date": "2023-10-01",
                "Location": "Test Location",
                "Description": "This is a test event.",
            },
        )
    )
    respx_mock.get("https://tix.blueroom.org.au/api/v1/Items/Browse").mock(
        Response(
            status_code=200,
            json={
                "Items": [
                    {
                        "Id": 1,
                        "Name": "Test Item 1",
                        "Price": 10.0,
                        "Hash": "abc123",
                        "DescriptionBrief": "A brief description of Test Item 1.",
                        "URL": "/Event/1/2023-10-01",
                        "Status": 0,
                        "VenueName": "Test Venue",
                        "Runtime": 120,
                        "DateTime": "2023-10-01T00:00:00Z",
                    }
                ]
            },
        )
    )

    data = uvloop.run(
        process_domain(
            "blueroom.org.au", datetime(2023, 10, 1, 0, 0, 0, tzinfo=timezone.utc)
        )
    )

    snapshot.assert_match(
        data.model_dump_json(indent=2),
    )
