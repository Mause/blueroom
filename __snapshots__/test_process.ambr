# serializer version: 1
# name: test_main
  '''
  {
    "updated_at": "2025-08-23T15:40:42.142446+08:00",
    "events": [
      {
        "item_hash": "abc123",
        "title": "Test Item 1",
        "url": "https://blueroom.org.au/event/1/",
        "html_desc": null,
        "desc": "A brief description of Test Item 1.",
        "dates": [
          {
            "start": "2023-10-01T00:00:00+08:00",
            "end": "2023-10-01T02:00:00+08:00",
            "venue": "Test Venue",
            "status": 0,
            "url": "https://tix.blueroom.org.au/Event/1/2023-10-01"
          }
        ]
      }
    ]
  }
  '''
# ---
